이미지 합성곱

-이미지는 2차원 -> 필터도 2차원 -> 현재 화소와 주위화소 배열을 각 요소 끼리 모두 곱하고 더함.  
-Zero padding : 바깥화소까지 처리해주기 위해 보통 0이나 1로 패딩함.  
-Pooling  
    -Max Pooling : 가장 큰 화소값 선택
    -Mean Pooling : 평균값 선택  
    -Min Pooling : 최솟값 선택
    -Overlapped Pooling : 

CNN 예시.
Convolution -> Pooling -> Convolution -> Pooling -> Fully Connected
Feature Extraction----------------------------------Classification

One-Hot-Encoding
각 카테고리 값을 이진 벡터로 표현.

#Conv2D는 2차원 이미지 데이터를 처리하기 위한 **합성곱 계층 (Convolutional Layer)**
-> 필터를 이용해 이미지 스캔
    •	filters: 출력 feature map의 수 (즉, 뉴런의 수)
	•	kernel_size: 필터의 크기 (예: (3,3)은 3x3 필터)
	•	activation: 활성화 함수 (relu, sigmoid 등)
	•	input_shape: (높이, 너비, 채널 수)


In [ ]:
from tensorflow.keras import Sequential, layers
model = Sequential([
    layers.Input(shape=(28,28,1)), #인풋값 정의 3차원은 RGB값인데 이건 흓백.
    #뉴런 하나당 필터만큼의 가중치, 편향
    #-> 3x3x32 + 32 = 128개의 파라미터가 존재함.
    layers.Conv2D(32, (3,3), activation='relu'), #컨볼루션레이어, 32개
    layers.MaxPooling2D(), # pool_size=(2,2)가 기본값
    layers.Dropout(0.1),    #10퍼센트 뉴런 드롭아웃
    layers.Conv2D(64, 3, activation='relu'), #컨볼루션레이어, 64개
    layers.MaxPooling2D(), #Max 풀링
    layers.Dropout(0.25), #25퍼센트 드롭아웃
    layers.Flatten(), #1차원으로 바꾸고
    layers.Dense(128, activation='relu'),#Dense 레이어 128개
    layers.Dropout(0.5), #드롭아웃 50퍼센트
    layers.Dense(10, activation='softmax') #소프트맥스로 출력
    ])
model.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_2 (Conv2D)               │ (None, 26, 26, 32)     │           320 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_2 (MaxPooling2D)  │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 13, 13, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_3 (Conv2D)               │ (None, 11, 11, 64)     │        18,496 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_3 (MaxPooling2D)  │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_4 (Dropout)             │ (None, 5, 5, 64)       │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_1 (Flatten)             │ (None, 1600)           │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 128)            │       204,928 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_5 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_3 (Dense)                 │ (None, 10)             │         1,290 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 225,034 (879.04 KB)

 Trainable params: 225,034 (879.04 KB)

 Non-trainable params: 0 (0.00 B)

| 레이어 이름            | 출력 크기 (Output Shape) | 파라미터 수 (Param #)         | 설명 |
|------------------------|---------------------------|-------------------------------|------|
| Input                  | (28, 28, 1)               | 0                             | 입력 이미지 (28x28, 흑백) |
| Conv2D (1)             | (26, 26, 32)              | 320                           | (3×3×1)×32 + 32 = 320 |
| MaxPooling2D (1)       | (13, 13, 32)              | 0                             | 다운샘플링만 수행 |
| Dropout (0.1)          | (13, 13, 32)              | 0                             | 10% 뉴런 무작위 비활성화 |
| Conv2D (2)             | (11, 11, 64)              | 18,496                        | (3×3×32)×64 + 64 = 18,496 |
| MaxPooling2D (2)       | (5, 5, 64)                | 0                             | 다운샘플링만 수행 |
| Dropout (0.25)         | (5, 5, 64)                | 0                             | 25% 드롭아웃 |
| Flatten                | (1600,)                   | 0                             | 5×5×64 = 1600개로 펼침 |
| Dense (1)              | (128,)                    | 204,928                       | 1600×128 + 128 = 204,928 |
| Dropout (0.5)          | (128,)                    | 0                             | 50% 드롭아웃 |
| Dense (2 - Output)     | (10,)                     | 1,290                         | 128×10 + 10 = 1,290 |

**총 파라미터 수**: 225,034개

In [4]:
model.compile(loss='sparse_categorical_crossentropy' ,optimizer='adam', metrics=['accuracy'])

In [6]:
from tensorflow.keras.datasets import mnist
(X_train, y_train), (X_test, y_test) = mnist.load_data()
X_train = X_train.reshape(-1, 28, 28, 1) / 255.0
X_test = X_test.reshape(-1, 28, 28, 1) / 255.0

In [7]:
model.fit(X_train, y_train, batch_size=128, epochs=10, verbose=1)

Epoch 1/10


2025-05-19 20:55:20.600809: I tensorflow/core/grappler/optimizers/custom_graph_optimizer_registry.cc:117] Plugin optimizer for device_type GPU is enabled.


469/469 ━━━━━━━━━━━━━━━━━━━━ 11s 19ms/step - accuracy: 0.7757 - loss: 0.6831
Epoch 2/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9573 - loss: 0.1431
Epoch 3/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9669 - loss: 0.1200
Epoch 4/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9681 - loss: 0.1214
Epoch 5/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9708 - loss: 0.1218
Epoch 6/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9728 - loss: 0.1148
Epoch 7/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9700 - loss: 0.1323
Epoch 8/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9721 - loss: 0.1392
Epoch 9/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9737 - loss: 0.1395
Epoch 10/10
469/469 ━━━━━━━━━━━━━━━━━━━━ 9s 19ms/step - accuracy: 0.9716 - loss: 0.1445


In [9]:
loss, accuracy = model.evaluate(X_test, y_test, verbose=0)
print(f'Test loss: {loss}, Test accuracy: {accuracy}')

Test loss: 0.05452838912606239, Test accuracy: 0.9896000027656555


LeNet : 최초의 CNN
    convolution - subsampling(average pooling) - convolution - subsampling - full connection - Gaussian connection

In [ ]:
class LeNet:
    @staticmethod
    def build(input_shape=(32,32,1), activation='sigmoid'):
        model = Sequential()
        model.add(Input(shape=input_shape))